wrappers/kv_quorum.go

```diff
@@ -0,0 +1,385 @@
+}
+
+// Version returns the supported version of the mem implementation
+func Version(url string, kvdbOptions map[string]string) (string, error) {
+	return kvdb.MemVersion1, nil
+}
+
+func (k *kvQuorumCheckFilter) String() string {
+	return k.kv.String()
+}
+
+func (k *kvQuorumCheckFilter) Capabilities() int {
+	return k.kv.Capabilities()
+}
+
+func (k *kvQuorumCheckFilter) Get(key string) (*kvdb.KVPair, error) {
+	if k.inQuorum() {
+		return k.kv.Get(key)
+	} else {
+		return nil, kvdb.ErrNoQuorum
+	}
+}
+
+func (k *kvQuorumCheckFilter) Snapshot(prefixes []string, consistent bool) (kvdb.Kvdb, uint64, error) {
+	if k.inQuorum() {
+		return k.kv.Snapshot(prefixes, consistent)
+	} else {
+		return nil, 0, kvdb.ErrNoQuorum
+	}
+}
+
+func (k *kvQuorumCheckFilter) Put(
+	key string,
+	value interface{},
+	ttl uint64,
+) (*kvdb.KVPair, error) {
+	if k.inQuorum() {
+		return k.kv.Put(key, value, ttl)
+	} else {
+		return nil, kvdb.ErrNoQuorum
+	}
+}
+
+func (k *kvQuorumCheckFilter) GetVal(key string, v interface{}) (*kvdb.KVPair, error) {
+	if k.inQuorum() {
+		return k.kv.GetVal(key, v)
+	} else {
+		return nil, kvdb.ErrNoQuorum
+	}
+}
+
+func (k *kvQuorumCheckFilter) Create(
+	key string,
+	value interface{},
+	ttl uint64,
+) (*kvdb.KVPair, error) {
+	if k.inQuorum() {
+		return k.kv.Create(key, value, ttl)
+	} else {
+		return nil, kvdb.ErrNoQuorum
+	}
+}
+
+func (k *kvQuorumCheckFilter) Update(
+	key string,
+	value interface{},
+	ttl uint64,
+) (*kvdb.KVPair, error) {
+	if k.inQuorum() {
+		return k.kv.Update(key, value, ttl)
+	} else {
+		return nil, kvdb.ErrNoQuorum
+	}
+}
+
+func (k *kvQuorumCheckFilter) Enumerate(prefix string) (kvdb.KVPairs, error) {
+	if k.inQuorum() {
+		return k.kv.Enumerate(prefix)
+	} else {
+		return nil, kvdb.ErrNoQuorum
+	}
+}
+
+func (k *kvQuorumCheckFilter) Delete(key string) (*kvdb.KVPair, error) {
+	if k.inQuorum() {
+		return k.kv.Delete(key)
+	} else {
+		return nil, kvdb.ErrNoQuorum
+	}
+}
+
+func (k *kvQuorumCheckFilter) DeleteTree(prefix string) error {
+	if k.inQuorum() {
+		return k.kv.DeleteTree(prefix)
+	} else {
+		return kvdb.ErrNoQuorum
+	}
+}
+
+func (k *kvQuorumCheckFilter) Keys(prefix, sep string) ([]string, error) {
+	if k.inQuorum() {
+		return k.kv.Keys(prefix, sep)
+	} else {
+		return nil, kvdb.ErrNoQuorum
+	}
+}
+
+func (k *kvQuorumCheckFilter) CompareAndSet(
+	kvp *kvdb.KVPair,
+	flags kvdb.KVFlags,
+	prevValue []byte,
+) (*kvdb.KVPair, error) {
+	if k.inQuorum() {
+		return k.kv.CompareAndSet(kvp, flags, prevValue)
+	} else {
+		return nil, kvdb.ErrNoQuorum
+	}
+}
+
+func (k *kvQuorumCheckFilter) CompareAndDelete(
+	kvp *kvdb.KVPair,
+	flags kvdb.KVFlags,
+) (*kvdb.KVPair, error) {
+	if k.inQuorum() {
+		return k.kv.CompareAndDelete(kvp, flags)
+	} else {
+		return nil, kvdb.ErrNoQuorum
+	}
+}
+
+func (k *kvQuorumCheckFilter) WatchKey(
+	key string,
+	waitIndex uint64,
+	opaque interface{},
+	cb kvdb.WatchCB,
+) error {
+	return k.kv.WatchKey(key, waitIndex, opaque, cb)
```

Comment:
Why not do the same for watch? The callers might retry subscribing a watch if connectivity to etcd is lost.

Reply:
watch is the way to determine if kvdb connection is restored.